tests/test_beamtime.py

```diff
@@ -0,0 +1,322 @@
+        self.assertEqual(apsbss.POLL_INTERVAL, 0.01)
+        self.assertEqual(
+            apsbss.DM_APS_DB_WEB_SERVICE_URL,
+            "https://xraydtn01.xray.aps.anl.gov:11236")
+        self.assertIsNotNone(apsbss.api_bss)
+        self.assertIsNotNone(apsbss.api_esaf)
+
+    def test_iso2datetime(self):
+        self.assertEqual(
+            apsbss.iso2datetime("2020-06-30 12:31:45.067890"),
+            datetime.datetime(2020, 6, 30, 12, 31, 45, 67890)
+        )
+
+    def test_not_at_aps(self):
+        self.assertTrue(True)   # test something
+        if using_APS_workstation():
+            return
+
+        # do not try to test for fails using dm package, it has no timeout
+
+    def test_only_at_aps(self):
+        self.assertTrue(True)   # test something
+        if not using_APS_workstation():
+            return
+
+        runs = apsbss.listAllRuns()
+        self.assertGreater(len(runs), 1)
+        self.assertEqual(apsbss.getCurrentCycle(), runs[-1])
+        self.assertEqual(apsbss.listRecentRuns()[0], runs[-1])
+
+        self.assertGreater(len(apsbss.listAllBeamlines()), 1)
+
+        # TODO: test the other functions
+        # getCurrentEsafs
+        # getCurrentInfo
+        # getCurrentProposals
+        # getEsaf
+        # getProposal
+        # class DmRecordNotFound(Exception): ...
+        # class EsafNotFound(DmRecordNotFound): ...
+        # class ProposalNotFound(DmRecordNotFound): ...
+
+    def test_printColumns(self):
+        from tests.common import Capture_stdout
+        with Capture_stdout() as received:
+            apsbss.printColumns("1 2 3 4 5 6".split(), numColumns=3, width=3)
+        self.assertEqual(len(received), 2)
+        self.assertEqual(received[0], "1  3  5  ")
+        self.assertEqual(received[1], "2  4  6  ")
+
+        source = "0123456789"
+        self.assertEqual(apsbss.trim(source), source)
+        self.assertNotEqual(apsbss.trim(source, length=8), source)
+        self.assertEqual(apsbss.trim(source, length=8), "01234...")
+
+
+class Test_EPICS(unittest.TestCase):
+
+    def setUp(self):
+        self.bss = None
+        self.manager = os.path.abspath(os.path.join(
+            os.path.dirname(apsbss.__file__),
+            "manage_ioc.sh"
+        ))
+        self.ioc_name = "test_apsbss"
+        cmd = f"{self.manager} restart {self.ioc_name} {BSS_TEST_IOC_PREFIX}"
+        self.ioc_process = subprocess.Popen(
+            cmd.encode().split(),
+            stdin=subprocess.PIPE,
+            stdout=subprocess.PIPE,
+            stderr=subprocess.PIPE,
+            shell=False)
+        time.sleep(.5)   # allow the IOC to start
+
+    def tearDown(self):
+        if self.bss is not None:
+            self.bss.destroy()
+            self.bss = None
+        if self.ioc_process is not None:
+            self.ioc_process = None
+            cmd = f"{self.manager} stop {self.ioc_name} {BSS_TEST_IOC_PREFIX}"
+            session = subprocess.Popen(
```

Comment:
Codacy found an issue: [local variable 'session' is assigned to but never used (F841)](https://app.codacy.com/gh/BCDA-APS/apstools/pullRequest?prid=5764884)